### PR TITLE
[REG2.068.1-b2] Issue 15017 - assigning a Variant to be value in a hashmap

### PR DESCRIPTION
**Pull request by 9rnsr**
https://issues.dlang.org/show_bug.cgi?id=15017

Request TypeInfo generation where the corresponding glue layer code needs it.

### CODE REVIEW

File: src/expression.c [hunk 1]
```diff
@@ -9571,7 +9573,10 @@ Expression *DeleteExp::semantic(Scope *sc)
                 FuncDeclaration *fd = sd->dtor;
 
                 if (!f)
+                {
+                    semanticTypeInfo(sc, ts);
```
Comment (MartinNowak) [inline src/expression.c]: These implicit dependencies are really the worst maintenance wise?
Is there any way we can do this in the glue code where the TypeInfo is actually requested?

Comment (9rnsr) [inline src/expression.c]: Unfortunately it's impossible. Currently a struct TypeInfo may refer to template member functions (`opEquals` and `opCmp`), and their mangled name is affected by the attribute inference result. So we should guarantee to run those semantic3 pass before the reach to glue layer. That's why the `semanticTypeInfo` call is necessary in each places.

**Comment by 9rnsr** [comment]
Thanks!

File: src/expression.c (hunk 1 above)
Comment (rainers) [inline src/expression.c]: I think the compiler needs a larger refactoring: the transformations/lowerings made in the glue layer must happen while semantic analysis is still possible. That would obsolete all this guessing what typeinfo will be needed during code generation, sometimes translated to something else making the typeinfo unnecessary. Without that change templated library implementation of associative and dynamic arrays are unlikely to produce good inlining results.

My current idea regarding such a refactoring would be to do the lowerings in a "semantic4" step, that starts once semantic3 is complete, but can still go back to the earlier stages. That way, user code will produce error messages as now, but low-level library code might show the internal names instead.

Comment (MartinNowak) [inline src/expression.c]: We already do lower things like to reach in earlier semantic steps so it should be possible for other constructs as well.

**Pull request merged**